File: rd-kt/rd-framework/src/jvmTest/kotlin/com/jetbrains/rd/framework/test/cases/wire/SocketWireTest.kt
package com.jetbrains.rd.framework.test.cases.wire

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.static
import com.jetbrains.rd.framework.impl.RdOptionalProperty
import com.jetbrains.rd.framework.impl.RdSignal
import com.jetbrains.rd.framework.test.util.TestBase
import com.jetbrains.rd.framework.test.util.TestScheduler
import com.jetbrains.rd.framework.util.NetUtils
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import com.jetbrains.rd.util.spinUntil
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import java.net.InetAddress
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.TimeoutException


class SocketWireTest : TestBase() {

    private fun <T : Any> RdOptionalProperty<T>.waitAndAssert(expected: T, prev: T? = null) {
        val start = System.currentTimeMillis()
        while ((System.currentTimeMillis() - start) < timeoutToWaitConditionMs && valueOrNull != expected) Thread.sleep(100)

        if (valueOrNull == prev) throw TimeoutException("Timeout $timeoutToWaitConditionMs ms while waiting value '$expected'")
        assertEquals(expected, valueOrNull)
    }

    companion object {
        internal const val top = "top"

        internal fun server(lifetime: Lifetime, port: Int? = null): Protocol {
            return Protocol("Server", Serializers(), Identities(IdKind.Server), TestScheduler,
                SocketWire.Server(lifetime, TestScheduler, port, "TestServer"), lifetime
            )
        }


        internal fun client(lifetime: Lifetime, serverProtocol: Protocol): Protocol {
            return Protocol("Client", Serializers(), Identities(IdKind.Client), TestScheduler,
                SocketWire.Client(lifetime,
                    TestScheduler, (serverProtocol.wire as SocketWire.Server).port, "TestClient"), lifetime
            )
        }

        internal fun client(lifetime: Lifetime, port: Int): Protocol {
            return Protocol("Client", Serializers(), Identities(IdKind.Client), TestScheduler,
                SocketWire.Client(lifetime, TestScheduler, port, "TestClient"), lifetime
            )
        }
    }

    lateinit var socketLifetime: Lifetime

    @BeforeEach
    fun setUp() {
        socketLifetime = lifetime.createNested().lifetime
//        ConsoleLoggerFactory.minLevelToLog = LogLevel.Trace
    }

    @Test
    fun TestBasicRun() {
        val serverProtocol = server(socketLifetime)
        val clientProtocol = client(socketLifetime, serverProtocol)

        val sp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, serverProtocol, top) }
        val cp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, clientProtocol, top) }

        cp.set(1)
        sp.waitAndAssert(1)

        sp.set(2)
        cp.waitAndAssert(2, 1)
    }

    @Test
    fun TestOrdering() {
        val serverProtocol = server(socketLifetime)
        val clientProtocol = client(socketLifetime, serverProtocol)

        val sp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, serverProtocol, top) }
        val cp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, clientProtocol, top) }

        val log = ConcurrentLinkedQueue<Int>()
        sp.advise(lifetime) { log.add(it) }
        cp.set(1)
        cp.set(2)
        cp.set(3)
        cp.set(4)
        cp.set(5)

        while (log.size < 5) Thread.sleep(100)
        assertEquals(listOf(1, 2, 3, 4, 5), log.toList())
    }


    @Disabled
    @Test
    fun TestDisconnect() {
        val serverProtocol = server(socketLifetime)
        val clientProtocol = client(socketLifetime, serverProtocol)

        val sp = RdSignal<Int>().static(1).apply { bind(lifetime, serverProtocol, top) }
        val cp = RdSignal<Int>().static(1).apply { bind(lifetime, clientProtocol, top) }

        val log = mutableListOf<Int>()
        sp.advise(socketLifetime) { log.add(it) }

        cp.fire(1)
        cp.fire(2)

        spinUntil { log.size == 2 }
        assertEquals(listOf(1, 2), log)

        tryCloseConnection(clientProtocol)
        cp.fire(3)
        cp.fire(4)

        spinUntil { log.size == 4 }
        assertEquals(listOf(1, 2, 3, 4), log)


        cp.fire(5)
        tryCloseConnection(serverProtocol)
        cp.fire(6)
        spinUntil { log.size == 6 }
        assertEquals(listOf(1, 2, 3, 4, 5, 6), log)

    }

    private fun tryCloseConnection(protocol: Protocol) {

        val wire = protocol.wire as SocketWire.Base
        if (spinUntil(50) { wire.connected.value }) {
            //if wire is not connected should not ask socket.close because it could lead to IOException
            // 	at java.net.Socket.getOutputStream(Socket.java:951)
            //	at com.jetbrains.rd.framework.SocketWire$Base$1.invoke(SocketWire.kt:102)
            wire.socketProvider.valueOrNull?.close()
        }
    }


    @Test
    fun TestDdos() {
        val serverProtocol = server(socketLifetime)
        val clientProtocol = client(socketLifetime, serverProtocol)

        val sp = RdSignal<Int>().static(1).apply { bind(lifetime, serverProtocol, top) }
        val cp = RdSignal<Int>().static(1).apply { bind(lifetime, clientProtocol, top) }

        var count = 0
        sp.advise(socketLifetime) {
            assertEquals(count + 1, it)
            ++count
        }
        val C = 50
        for (i in 1..C) {
            cp.fire(i)
            if (i == C / 2) {
                tryCloseConnection(serverProtocol)
            }
        }

        spinUntil { count == C }
    }


    @Test
    fun TestBigBuffer() {
        val serverProtocol = server(socketLifetime)
        val clientProtocol = client(socketLifetime, serverProtocol)

        val sp = RdOptionalProperty<String>().static(1).apply { bind(lifetime, serverProtocol, top) }
        val cp = RdOptionalProperty<String>().static(1).apply { bind(lifetime, clientProtocol, top) }

        cp.set("1")
        sp.waitAndAssert("1")

        sp.set("".padStart(100000, '3'))
        cp.waitAndAssert("".padStart(100000, '3'), "1")
    }


    @Test
    fun TestRunWithSlowpokeServer() {

        val port = NetUtils.findFreePort(0)
        val clientProtocol = client(socketLifetime, port)


        val cp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, clientProtocol, top) }

        cp.set(1)

        Thread.sleep(2000)

        val serverProtocol = server(socketLifetime, port)
        val sp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, serverProtocol, top) }

        val prev = sp.valueOrNull
        cp.set(4)
        sp.waitAndAssert(4, prev)
    }

    @Test
    fun TestServerWithoutClient() {
        server(socketLifetime)
    }

    @Test
    fun TestServerWithoutClientWithDelay() {
        server(socketLifetime)
        Thread.sleep(100)
    }

    @Test
    fun TestServerWithoutClientWithDelayAndMessages() {
        val protocol = server(socketLifetime)
        Thread.sleep(100)
        val sp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, protocol, top) }

        sp.set(1)
        sp.set(2)
        Thread.sleep(50)
    }

    @Test
    fun TestClientWithoutServer() {
        client(socketLifetime, NetUtils.findFreePort(0))
    }

    @Test
    fun TestClientWithoutServerWithDelay() {
        client(socketLifetime, NetUtils.findFreePort(0))
        Thread.sleep(100)
    }

    @Test
    fun TestClientWithoutServerWithDelayAndMessages() {
        val clientProtocol = client(socketLifetime, NetUtils.findFreePort(0))

        val cp = RdOptionalProperty<Int>().static(1).apply { bind(lifetime, clientProtocol, top) }

        cp.set(1)
        cp.set(2)
        Thread.sleep(50)
    }

//    @Test
//    fun testReentrantWrites() {
//        val serverProtocol = server(socketLifetime)
//        val clientProtocol = client(socketLifetime, serverProtocol)
//
//        serverProtocol.serializers.register(InterningNestedTestStringModel)
//        clientProtocol.serializers.register(InterningNestedTestStringModel)
//
//        val sp = PropertyHolderWithInternRoot(RdOptionalProperty<InterningNestedTestStringModel>().static(1), serverProtocol.serializationContext)
//        val cp = PropertyHolderWithInternRoot(RdOptionalProperty<InterningNestedTestStringModel>().static(1), clientProtocol.serializationContext)
//
//        sp.mySerializationContext = sp.mySerializationContext.withInternRootsHere(sp, "Test")
//        cp.mySerializationContext = cp.mySerializationContext.withInternRootsHere(cp, "Test")
//
//        sp.bind(lifetime, serverProtocol, "top")
//        cp.bind(lifetime, clientProtocol, "top")
//
//        val modelA = InterningNestedTestStringModel("A", InterningNestedTestStringModel("B", InterningNestedTestStringModel("C", null)))
//        cp.property.set(modelA)
//        sp.property.waitAndAssert(modelA)
//
//        val modelB = InterningNestedTestStringModel("D", InterningNestedTestStringModel("E", InterningNestedTestStringModel("F", null)))
//        sp.property.set(modelB)
//        cp.property.waitAndAssert(modelB, modelA)
//    }


    @Test
    fun testRemoteSocket() {
        val serverSocket = SocketWire.Server(lifetime, TestScheduler, 0, allowRemoteConnections = true)
        val clientSocket = SocketWire.Client(lifetime, TestScheduler, serverSocket.port, hostAddress = InetAddress.getLocalHost())

        assertTrue(spinUntil(60000L) { clientSocket.connected.value })
    }



    @Test
    fun testSocketFactory() {

        val sLifetime = LifetimeDefinition()
        val factory = SocketWire.ServerFactory(sLifetime, TestScheduler, 0)

        val lf1 = LifetimeDefinition()
        SocketWire.Client(lf1, TestScheduler, factory.localPort)

        spinUntil { factory.size == 1 }

        val lf2 = LifetimeDefinition()
        SocketWire.Client(lf2, TestScheduler, factory.localPort)

        spinUntil { factory.size == 2 }

        lf1.terminate()
        spinUntil { factory.size == 1 }

        sLifetime.terminate()
        spinUntil { factory.size == 0 }
    }

    @ParameterizedTest
    @ValueSource(booleans = [true, false])
    fun testPacketLoss(isClientToServer: Boolean) {
        Lifetime.using { lifetime ->
            val serverProtocol = server(lifetime)
            val serverWire = serverProtocol.wire

            val proxy = SocketProxy("TestProxy", lifetime, serverProtocol)

            proxy.start()

            val clientProtocol = client(lifetime, proxy.port)
            val clientWire = clientProtocol.wire

            Thread.sleep(100)

            if (isClientToServer)
                proxy.stopClientToServerMessaging()
            else
                proxy.stopServerToClientMessaging()

            val detectionTimeoutMs = clientProtocol.wire.heartbeatIntervalMs * (SocketWire.maximumHeartbeatDelay + 3)

            Thread.sleep(detectionTimeoutMs)

            assertTrue(serverWire.connected.value)
            assertTrue(clientWire.connected.value)

            assertFalse(serverWire.heartbeatAlive.value)
            assertFalse(clientWire.heartbeatAlive.value)

            if (isClientToServer)
                proxy.startClientToServerMessaging()
            else
                proxy.startServerToClientMessaging()

            Thread.sleep(detectionTimeoutMs)

            assertTrue(serverWire.connected.value)
            assertTrue(clientWire.connected.value)

            assertTrue(serverWire.heartbeatAlive.value)
            assertTrue(clientWire.heartbeatAlive.value)
        }
    }


//    @BeforeClass
//    fun beforeClass() {
//         setupLogHandler {
//        if (it.getLevel() == Level.ERROR) {
//            System.err.println(it.message)
//            it.throwableInformation?.throwable?.printStackTrace()
//        }
//         }
//    }
//
//    private fun setupLogHandler(name: String = "default", action: (LoggingEvent) -> Unit) {
//         val rootLogger = org.apache.log4j.Logger.getRootLogger()
//         rootLogger.removeAppender("default")
//         rootLogger.addAppender(object : AppenderSkeleton() {
//        init {
//            setName(name)
//        }
//
//        override fun append(event: LoggingEvent) {
//            action(event)
//        }
//
//        override fun close() {}
//
//        override fun requiresLayout(): Boolean {
//            return false
//        }
//         })
//    }
}

File: rd-net/Test.RdFramework/SocketWireTest.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using JetBrains.Collections.Viewable;
using JetBrains.Core;
using JetBrains.Diagnostics;
using JetBrains.Diagnostics.Internal;
using JetBrains.Lifetimes;
using JetBrains.Rd;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Threading;
using NUnit.Framework;
using Test.Lifetimes;

namespace Test.RdFramework
{
  [TestFixture]
  public class SocketWireTest : LifetimesTestBase
  {
    internal static TimeSpan DefaultTimeout = TimeSpan.FromMilliseconds(100);

    internal const string Top = "top";
    private void WaitAndAssert<T>(RdProperty<T> property, T expected, T prev)
    {
      WaitAndAssert(property, expected, new Maybe<T>(prev));
    }


    private void WaitAndAssert<T>(RdProperty<T> property, T expected, Maybe<T> prev = default(Maybe<T>))
    {
      var start = Environment.TickCount;
      const int timeout = 5000;
      while (Environment.TickCount - start < timeout && property.Maybe == prev) Thread.Sleep(10);
      if (property.Maybe == prev)
        throw new TimeoutException($"Timeout {timeout} ms while waiting for value '{expected}'");
      Assert.AreEqual(expected, property.Value);
    }


    static int FindFreePort()
    {
      TcpListener l = new TcpListener(IPAddress.Loopback, 0);
      l.Start();
      int port = ((IPEndPoint) l.LocalEndpoint).Port;
      l.Stop();
      return port;
    }


    internal static IProtocol Server(Lifetime lifetime, int? port = null)
    {
      var id = "TestServer";
      var server = new SocketWire.Server(lifetime, SynchronousScheduler.Instance, new IPEndPoint(IPAddress.Loopback, port ?? 0), id);
      return new Protocol(id, new Serializers(), new SequentialIdentities(IdKind.Server), SynchronousScheduler.Instance, server, lifetime);
    }

    internal static IProtocol Client(Lifetime lifetime, int port)
    {
      var id = "TestClient";
      var client = new SocketWire.Client(lifetime, SynchronousScheduler.Instance, port, id);
      return new Protocol(id, new Serializers(), new SequentialIdentities(IdKind.Server), SynchronousScheduler.Instance, client, lifetime);
    }

    internal static IProtocol Client(Lifetime lifetime, IProtocol serverProtocol)
    {
      // ReSharper disable once PossibleNullReferenceException
      return Client(lifetime, (serverProtocol.Wire as SocketWire.Server).Port);
    }

    [Test]
    public void TestBasicRun()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        var serverProtocol = Server(lifetime);
        var clientProtocol = Client(lifetime, serverProtocol);

        var sp = NewRdProperty<int>().Static(1);
        sp.BindTopLevel(lifetime, serverProtocol, Top);
        var cp = NewRdProperty<int>().Static(1);
        cp.BindTopLevel(lifetime, clientProtocol, Top);

        cp.SetValue(1);
        WaitAndAssert(sp, 1);
      });
    }


    [Test]
    public void TestOrdering()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        var serverProtocol = Server(lifetime);
        var clientProtocol = Client(lifetime, serverProtocol);

        var sp = NewRdProperty<int>().Static(1);
        sp.BindTopLevel(lifetime, serverProtocol, Top);
        var cp = NewRdProperty<int>().Static(1);
        cp.BindTopLevel(lifetime, clientProtocol, Top);

        var log = new List<int>();
        sp.Advise(lifetime, it => log.Add(it));
        sp.SetValue(1);
        sp.SetValue(2);
        sp.SetValue(3);
        sp.SetValue(4);
        sp.SetValue(5);

        while (log.Count < 5) Thread.Sleep(10);
        CollectionAssert.AreEqual(new[] {1, 2, 3, 4, 5}, log);
      });
    }


    [Test]
    public void TestBigBuffer()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        var serverProtocol = Server(lifetime);
        var clientProtocol = Client(lifetime, serverProtocol);

        var sp = NewRdProperty<string>().Static(1);
        sp.BindTopLevel(lifetime, serverProtocol, Top);
        var cp = NewRdProperty<string>().Static(1);
        cp.BindTopLevel(lifetime, clientProtocol, Top);

        cp.SetValue("1");
        WaitAndAssert(sp, "1");

        sp.SetValue(new string('a', 100000));
        WaitAndAssert(cp, new string('a', 100000), "1");

        cp.SetValue("a");
        WaitAndAssert(sp, "a", new string('a', 100000));

        cp.SetValue("ab");
        WaitAndAssert(sp, "ab", "a");

        cp.SetValue("abc");
        WaitAndAssert(sp, "abc", "ab");
      });
    }


    [Test]
    public void TestRunWithSlowpokeServer()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);

        var port = FindFreePort();
        var clientProtocol = Client(lifetime, port);

        var cp = NewRdProperty<int>().Static(1);
        cp.BindTopLevel(lifetime, clientProtocol, Top);
        cp.SetValue(1);

        Thread.Sleep(2000);
        var serverProtocol = Server(lifetime, port);
        var sp = NewRdProperty<int>().Static(1);
        sp.BindTopLevel(lifetime, serverProtocol, Top);

        var prev = sp.Maybe;


        cp.SetValue(4);
        Thread.Sleep(200);
        WaitAndAssert(sp, 4, prev);
      });
    }


    [Test]
    [Timeout(5000)]
    public void TestServerWithoutClient()
    {
      Lifetime.Using(lifetime =>
      {
        WithLongTimeout(lifetime);
        SynchronousScheduler.Instance.SetActive(lifetime);
        Server(lifetime);
      });
    }

    [Test]
    public void TestServerWithoutClientWithDelay()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        Server(lifetime);
        Thread.Sleep(100);
      });
    }

    [Test]
    public void TestServerWithoutClientWithDelayAndMessages()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        var protocol = Server(lifetime);
        Thread.Sleep(100);
        var p = NewRdProperty<int>().Static(1);
        p.BindTopLevel(lifetime, protocol, Top);
        p.SetValue(1);
        p.SetValue(2);
        Thread.Sleep(50);
      });
    }


    [Test]
    [Timeout(5000)]
    public void TestClientWithoutServer()
    {
      Lifetime.Using(lifetime =>
      {
        WithLongTimeout(lifetime);
        SynchronousScheduler.Instance.SetActive(lifetime);
        Client(lifetime, FindFreePort());
      });
    }

    [Test]
    public void TestClientWithoutServerWithDelay()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        Client(lifetime, FindFreePort());
        Thread.Sleep(100);
      });
    }

    [Test]
    public void TestClientWithoutServerWithDelayAndMessages()
    {
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        var protocol = Client(lifetime, FindFreePort());
        Thread.Sleep(100);
        var p = NewRdProperty<int>().Static(1);
        p.BindTopLevel(lifetime, protocol, Top);
        p.SetValue(1);
        p.SetValue(2);
        Thread.Sleep(50);
      });
    }


    [Test, Ignore("https://github.com/JetBrains/rd/issues/69")]
    public void TestDisconnect() => TestDisconnectBase((list, i) => list.Add(i));

    [Test]
    public void TestDisconnect_AllowDuplicates() => TestDisconnectBase((list, i) =>
    {
      // values may be duplicated due to asynchronous acknowledgement
      if (list.LastOrDefault() < i)
        list.Add(i);
    });

    private void TestDisconnectBase(Action<List<int>, int> advise)
    {
      var timeout = TimeSpan.FromSeconds(1);
      
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);
        var serverProtocol = Server(lifetime);
        var clientProtocol = Client(lifetime, serverProtocol);

        var sp = NewRdSignal<int>().Static(1);
        sp.BindTopLevel(lifetime, serverProtocol, Top);

        var cp = NewRdSignal<int>().Static(1);
        cp.BindTopLevel(lifetime, clientProtocol, Top);

        var log = new List<int>();
        sp.Advise(lifetime, i => advise(log, i));

        cp.Fire(1);
        cp.Fire(2);
        Assert.True(SpinWaitEx.SpinUntil(timeout, () => log.Count == 2));
        Assert.AreEqual(new List<int> {1, 2}, log);

        CloseSocket(clientProtocol);
        cp.Fire(3);
        cp.Fire(4);

        Assert.True(SpinWaitEx.SpinUntil(timeout, () => log.Count == 4));
        Assert.AreEqual(new List<int> {1, 2, 3, 4}, log);

        CloseSocket(serverProtocol);
        cp.Fire(5);
        cp.Fire(6);

        Assert.True(SpinWaitEx.SpinUntil(timeout, () => log.Count == 6));
        Assert.AreEqual(new List<int> {1, 2, 3, 4, 5, 6}, log);
      });
    }

    [Test]
    public void TestReconnect()
    {
      Lifetime.Using(lifetime => 
        {
          SynchronousScheduler.Instance.SetActive(lifetime);
          var serverProtocol = Server(lifetime, null);
          
          var sp = NewRdProperty<int>().Static(1);
          sp.BindTopLevel(lifetime, serverProtocol, Top);
          sp.IsMaster = false;

          var wire = serverProtocol.Wire as SocketWire.Base;
          int clientCount = 0;
          wire.NotNull().Connected.WhenTrue(lifetime, _ =>
          {
            clientCount++;
          });
          
          Assert.AreEqual(0, clientCount);
          
          Lifetime.Using(lf =>
          {
            var clientProtocol = Client(lf, serverProtocol);
            var cp = NewRdProperty<int>().Static(1);
            cp.IsMaster = true;
            cp.BindTopLevel(lf, clientProtocol, Top);
            cp.SetValue(1);            
            WaitAndAssert(sp, 1);            
            Assert.AreEqual(1, clientCount);
          });

          
          Lifetime.Using(lf =>
          {
            sp = NewRdProperty<int>().Static(2);
            sp.BindTopLevel(lifetime, serverProtocol, Top);
            
            var clientProtocol = Client(lf, serverProtocol);
            var cp = NewRdProperty<int>().Static(2);
            cp.BindTopLevel(lf, clientProtocol, Top);
            cp.SetValue(2);
            WaitAndAssert(sp, 2);
            Assert.AreEqual(2, clientCount);
          });                  
          
          
          Lifetime.Using(lf =>
          {                        
            var clientProtocol = Client(lf, serverProtocol);
            var cp = NewRdProperty<int>().Static(2);
            cp.BindTopLevel(lf, clientProtocol, Top);
            cp.SetValue(3);      
            WaitAndAssert(sp, 3, 2);
            Assert.AreEqual(3, clientCount);
          });      

        });
      
    }
    
    [TestCase(true)]
    [TestCase(false)]
    public void TestPacketLoss(bool isClientToServer)
    {
      using (Log.UsingLogFactory(new TextWriterLogFactory(Console.Out, LoggingLevel.TRACE)))
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);

        var serverProtocol = Server(lifetime);
        var serverWire = (SocketWire.Base) serverProtocol.Wire;

        var proxy = new SocketProxy("TestProxy", lifetime, serverProtocol);
        proxy.Start();

        var clientProtocol = Client(lifetime, proxy.Port);
        var clientWire = (SocketWire.Base) clientProtocol.Wire;

        Thread.Sleep(DefaultTimeout);

        if (isClientToServer)
          proxy.StopClientToServerMessaging();
        else
          proxy.StopServerToClientMessaging();

        var detectionTimeoutTicks = ((SocketWire.Base) clientProtocol.Wire).HeartBeatInterval.Ticks *
                                    (SocketWire.Base.MaximumHeartbeatDelay + 3);
        var detectionTimeout = TimeSpan.FromTicks(detectionTimeoutTicks);
          
        Thread.Sleep(detectionTimeout);

        Assert.IsTrue(serverWire.Connected.Value);
        Assert.IsTrue(clientWire.Connected.Value);
          
        Assert.IsFalse(serverWire.HeartbeatAlive.Value);
        Assert.IsFalse(clientWire.HeartbeatAlive.Value);

        if (isClientToServer)
          proxy.StartClientToServerMessaging();
        else
          proxy.StartServerToClientMessaging();

        Thread.Sleep(detectionTimeout);

        Assert.IsTrue(serverWire.Connected.Value);
        Assert.IsTrue(clientWire.Connected.Value);
          
        Assert.IsTrue(serverWire.HeartbeatAlive.Value);
        Assert.IsTrue(clientWire.HeartbeatAlive.Value);

      });
    }

    [Test]
    [Ignore("Not enough timeout to get the correct test")]
    public void TestStressHeartbeat()
    {
      // using (Log.UsingLogFactory(new TextWriterLogFactory(Console.Out, LoggingLevel.TRACE)))
      Lifetime.Using(lifetime =>
      {
        SynchronousScheduler.Instance.SetActive(lifetime);

        var interval = TimeSpan.FromMilliseconds(50);

        var serverProtocol = Server(lifetime);
        var serverWire = ((SocketWire.Base) serverProtocol.Wire).With(wire => wire.HeartBeatInterval = interval);

        var latency = TimeSpan.FromMilliseconds(40);
        var proxy = new SocketProxy("TestProxy", lifetime, serverProtocol) {Latency = latency};
        proxy.Start();

        var clientProtocol = Client(lifetime, proxy.Port);
        var clientWire = ((SocketWire.Base) clientProtocol.Wire).With(wire => wire.HeartBeatInterval = interval);

        Thread.Sleep(DefaultTimeout);

        serverWire.HeartbeatAlive.WhenFalse(lifetime, _ => Assert.Fail("Detected false disconnect on server side"));
        clientWire.HeartbeatAlive.WhenFalse(lifetime, _ => Assert.Fail("Detected false disconnect on client side"));

        Thread.Sleep(TimeSpan.FromSeconds(50));
      });
    }



    [Test]
    public void TestSocketFactory()
    {
      var sLifetime = new LifetimeDefinition();
      var factory = new SocketWire.ServerFactory(sLifetime.Lifetime, SynchronousScheduler.Instance);
      
      var lf1 = new LifetimeDefinition();
      new SocketWire.Client(lf1.Lifetime, SynchronousScheduler.Instance, factory.LocalPort);
      SpinWaitEx.SpinUntil(() => factory.Connected.Count == 1);
      
      var lf2 = new LifetimeDefinition();
      new SocketWire.Client(lf2.Lifetime, SynchronousScheduler.Instance, factory.LocalPort);
      SpinWaitEx.SpinUntil(() => factory.Connected.Count == 2);
      
      
      lf1.Terminate();
      SpinWaitEx.SpinUntil(() => factory.Connected.Count == 1);
      
      sLifetime.Terminate();
      SpinWaitEx.SpinUntil(() => factory.Connected.Count == 0);
    }


    private static void CloseSocket(IProtocol protocol)
    {
      if (!(protocol.Wire is SocketWire.Base socketWire))
      {
        Assert.Fail();
        return;
      }
      
      SocketWire.Base.CloseSocket(socketWire.Socket.NotNull());
    }

    private static void WithLongTimeout(Lifetime lifetime)
    {
      var oldValue = SocketWire.Base.TimeoutMs;
      lifetime.Bracket(() => SocketWire.Base.TimeoutMs = 100_000, () => SocketWire.Base.TimeoutMs = oldValue);
    }
  }
}